src/locale/keymap-util.c

```diff
@@ -780,3 +782,148 @@ int x11_convert_to_vconsole(Context *c) {
+}
+
+bool locale_gen_check_available(void) {
+#if HAVE_LOCALEGEN
+        if (access(LOCALEGEN_PATH, X_OK) < 0) {
+                if (errno != ENOENT)
+                        log_warning_errno(errno, "Unable to determine whether " LOCALEGEN_PATH " exists and is executable: %m");
+                return false;
+        }
+        if (access("/etc/locale.gen", F_OK) < 0) {
+                if (errno != ENOENT)
+                        log_warning_errno(errno, "Unable to determine whether /etc/locale.gen exists: %m");
+                return false;
+        }
+        return true;
+#else
+        return false;
+#endif
+}
+
+int locale_gen_enable_locale(const char *locale) {
+        _cleanup_fclose_ FILE *fr = NULL, *fw = NULL;
+        _cleanup_(unlink_and_freep) char *temp_path = NULL;
+        _cleanup_free_ char *locale_entry = NULL;
+        bool locale_enabled = false, first_line = false;
+        struct stat st;
+        int r;
+
+        if (isempty(locale))
+                return 0;
+
+        if (locale_encoding_is_utf8_or_unspecified(locale)) {
+                locale_entry = strjoin(locale, " UTF-8");
+                if (!locale_entry)
+                        return -ENOMEM;
+        } else
+                return -ENOEXEC; /* We do not process non-UTF-8 locale */
+
+        fr = fopen("/etc/locale.gen", "re");
+        if (!fr)
+                return -errno;
+
+        r = fopen_temporary("/etc/locale.gen", &fw, &temp_path);
+        if (r < 0)
+                return r;
+
+        /* apply mode of the original file to new file */
+        if (fstat(fileno(fr), &st) < 0)
+                return -errno;
+        (void) fchmod(fileno(fw), st.st_mode);
+
+        /* The config file ends with a line break, which we do not want to include before potentially appending a new locale
+         * instead of uncommenting an existing line. By prepending linebreaks, we can avoid buffering this file but can still write
+         * a nice config file without empty lines */
+        first_line = true;
+        for (;;) {
+                _cleanup_free_ char *line = NULL;
+                char *line_locale;
+
+                r = read_line(fr, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                if (locale_enabled) {
+                        /* Just complete writing the file if the new locale was already enabled */
+                        if (!first_line)
+                                fputc('\n', fw);
+                        fputs(line, fw);
+                        first_line = false;
+                        continue;
+                }
+
+                line = strstrip(line);
+                if (isempty(line)) {
+                        fputc('\n', fw);
+                        first_line = false;
+                        continue;
+                }
+
+                line_locale = line;
```

Comment:
hmm, actually, I think it would be wise to exit here early if the line already exists in the file uncommented, i.e. we should leave the existing file in place unmodified if there's no reason to update. i.e. add a quick checke somewhere here that if the line is already in perfect order you do a quick "return 0" here, so that we we indicate success even if we didn't do anything, leaving the original file in place

Reply:
Right, that makes sense - this event won't even be rare anymore, now that we call this function multiple times in a loop.